src/Microsoft.Extensions.DependencyInjection.Ordered/Ordered.cs

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+
+namespace Microsoft.Extensions.DependencyInjection.Ordered
+{
+    internal class Ordered<T>: IOrdered<T>, IDisposable
+    {
+        private readonly IServiceProvider _serviceProvider;
+        private readonly OrderedEnumerableServiceDescriptorContainer<T> _descriptorContainer;
+        private List<T> _values;
+        private List<IDisposable> _dispose;
+
+        public Ordered(IServiceProvider serviceProvider, OrderedEnumerableServiceDescriptorContainer<T> descriptorContainer)
+        {
+            _serviceProvider = serviceProvider;
+            _descriptorContainer = descriptorContainer;
+        }
+
+        private void EnsureValues()
+        {
+            lock (_descriptorContainer)
```

Comment:
Is there a reason to specifically lock on the container? If not, just use a lock object.